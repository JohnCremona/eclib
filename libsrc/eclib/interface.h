// interface.h: used to provide common interface for LiDIA or NTL
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
//  The following macro switches can be used: it is best to define them
//  in Makefiles rather than changing this file.
//  1. NTL_INTS     Use NTL for bigints (no bigrationals)
//  2. NTL_ALL      Use NTL also for bigfloats (RR) and bigcomplexes (CC)
//  3. (neither)    Defaults to NTL_ALL

#ifndef _INTERFACE_H_
#define _INTERFACE_H_

#ifdef NTL_ALL
#define NTL_INTS
#define MPFP
#endif

#include <limits>
#include <iostream>
#include <sstream>
#include <fstream>
#include <set>
#include <vector>
#include <map>
#include <algorithm>
//#include <numeric>
#include <ext/numeric>
#include <iterator>
using namespace std;
#include "templates.h"
#include <stdint.h>

#ifndef MININT
#define MININT numeric_limits<int>::min()
#endif
#ifndef MINLONG
#define MINLONG numeric_limits<long>::min()
#endif
#ifndef MAXINT
#define MAXINT numeric_limits<int>::max()
#endif
#ifndef MAXLONG
#define MAXLONG numeric_limits<long>::max()
#endif

// integers and rationals

// Some of the following were defined for compatibility with LiDIA, which is no longer supported

#include <NTL/ZZ.h>
#include <NTL/ZZXFactoring.h>
using namespace NTL;

#define bigint ZZ
//#define bigrational QQ  // not defined in NTL

#define BIGINT(val) to_ZZ(val)
inline bigint atoI(const char* s) {return to_ZZ(s);}

inline int odd(const int& a) {return a&1;}
inline int even(const int& a) {return !(a&1);}
inline int odd(const long& a) {return a&1;}
inline int even(const long& a) {return !(a&1);}

inline int is_zero(const bigint& x) {return IsZero(x);}
inline int is_positive(const bigint& x) {return sign(x)>0;}
inline int is_negative(const bigint& x) {return sign(x)<0;}
inline int is_one(const bigint& x) {return IsOne(x);}
inline int odd(const bigint& a) {return IsOdd(a);}
inline int even(const bigint& a) {return !IsOdd(a);}
inline void rshift(const bigint& a, long i, bigint& c) {RightShift(c,a,i);}
inline void lshift(const bigint& a, long i, bigint& c) {LeftShift(c,a,i);}
#ifdef setbit
#undef setbit
#endif
inline void setbit(bigint& a, int e) {SetBit(a,e);}
inline long lg(const bigint& x) {return NumBits(x)-1;}
inline int is_long(const bigint& a) {return (a<=MAXLONG)&&(a>=MINLONG);}
inline int is_int(const bigint& a) {return (a<=MAXINT)&&(a>=MININT);}

// The following are not in NTL & need defining
int I2int(const bigint& x);    // too long to inline
long I2long(const bigint& x);  // too long to inline
inline double I2double(const bigint& x) {return to_double(x);}
inline void longasI(long& a, const bigint& x) {a = I2long(x);}
inline void negate(bigint& a) {a=-a;}
inline void sqrt(bigint& a, const bigint& b) {SqrRoot(a,b);}
inline bigint sqrt(const bigint& a) {bigint b; sqrt(b,a); return b;}
inline void square(bigint& a, const bigint& b) {sqr(a,b);}
inline bigint gcd(const bigint& a, const bigint& b) {return GCD(a,b);}
inline bigint lcm(const bigint& a, const bigint& b)
 {if (IsZero(a) && IsZero(b)) return ZZ::zero(); else return a*(b/GCD(a,b));}

// In NTL add, sub, mul, div are defined with result in first place
inline void addx(const bigint& a, const bigint& b, bigint& c)  {add(c,a,b);}
inline void subx(const bigint& a, const bigint& b, bigint& c)  {sub(c,a,b);}
inline void divx(const bigint& a, const bigint& b, bigint& c)  {div(c,a,b);}
inline void mulx(const bigint& a, const bigint& b, bigint& c)  {mul(c,a,b);}
inline bigint pow(const bigint& a, long e)  {return power(a,e);}

//N.B. no power to bigint exponent in NTL
inline long jacobi(const bigint& a, const bigint& p)  {return Jacobi(a,p);}
inline void sqrt_mod_p(bigint & x, const bigint & a, const bigint & p)  
  {SqrRootMod(x,a,p); if(x>(p-x)) x= p-x;}
inline void power_mod(bigint& ans, const bigint& base, const bigint& expo, const bigint& m) 
 {PowerMod(ans,base,expo,m);}
inline void nearest(bigint& c, const bigint& a, const bigint& b) 
 {bigint a0=(a%b);  c = (a-a0)/b; if(2*a0>b) c+=1;}
inline bigint roundover(const bigint& a, const bigint& b)
 {bigint a0=(a%b); bigint c = (a-a0)/b; if(2*a0>b) c+=1; return c;}

#define bigint_mod_long(a,m) (a%m)


// Reals and Complexes

#ifdef NTL_ALL

#include <NTL/RR.h>
#define bigfloat RR
RR Pi();
RR Euler();
RR atan(const RR&);
RR asin(const RR&);
inline RR pow(const RR& a, int e)  {return power(a,e);}
inline RR pow(const RR& a, long e)  {return power(a,e);}
namespace NTL {
inline RR cosh(const RR& x) {return (exp(x)+exp(-x))/2;}
inline RR sinh(const RR& x) {return (exp(x)-exp(-x))/2;}
inline RR tan(const RR& x) {return sin(x)/cos(x);}
RR atan2(const RR&, const RR&);
inline int is_approx_zero(const RR& x)
//  {return abs(x)<power2_RR(2-RR::precision());}
{
  if (IsZero(x)) return 1;
  long n = x.exponent()+RR::precision()-1;
  if (n>=0) return 0;
  // cout<<"x="<<x<<", exponent="<<x.exponent()<<", mantissa="<<x.mantissa()<<", precision="<<RR::precision()<<endl;
  // cout<<"is_approx_zero() returns "<<(x.mantissa()<power2_ZZ(-n))<<endl;
  return abs(x.mantissa())<power2_ZZ(-n);}
} // namespace NTL
#include <complex>
typedef complex<RR> CC;
#define bigcomplex CC

inline void set_precision(long n)
  {RR::SetPrecision(long(n*3.33));RR::SetOutputPrecision(n);}
inline void set_bit_precision(long n)
  {RR::SetPrecision(n);}
inline void set_precision(const string prompt)
  {long n; cerr<<prompt<<": "; cin>>n; set_precision(n);}
inline long decimal_precision() {return long(RR::precision()*0.3);}
inline long bit_precision() {return RR::precision();}
inline int is_approx_zero(const bigcomplex& z)
  {return is_approx_zero(z.real())&&is_approx_zero(z.imag());}
inline RR to_bigfloat(const int& n) {return to_RR(n);}
inline RR to_bigfloat(const long& n) {return to_RR(n);}
inline RR to_bigfloat(const double& x) {return to_RR(x);}
inline RR I2bigfloat(const bigint& x) { return to_RR(x);}
inline int doublify(const bigfloat& x, double& d){ d=to_double(x); return 0;}
inline long longify(bigfloat x) {return to_long(x);}
inline int is_zero(bigfloat x) {return IsZero(x);}
inline int is_zero(bigcomplex z) {return IsZero(z.real()) && IsZero(z.imag());}
inline void Iasb(bigint& a, bigfloat x) {RoundToZZ(a,x);}
inline void Iasb(long& a, bigfloat x) {ZZ n; RoundToZZ(n,x); a=I2long(n);}
istream& operator>>(istream& is, CC& z);
inline CC pow(const CC& a, int e)  {return exp(to_RR(e)*log(a));}
inline CC pow(const CC& a, long e)  {return exp(to_RR(e)*log(a));}
inline CC pow(const CC& a, const RR& e)  {return exp(e*log(a));}


//////////////////////////////////////////////////////////////////
#else  // C doubles and libg++ Complexes
//////////////////////////////////////////////////////////////////

// reals

#define bigfloat double

inline long decimal_precision() {return 15;}
inline int is_zero(double x) {return fabs(x)<1e-15;}
inline int is_approx_zero(double x) {return fabs(x)<1e-10;}
inline void set_precision(long n) {cout.precision(n);}
inline void set_precision(const string prompt)  {cout.precision(15);}
#define Pi()    (3.1415926535897932384626433832795028841)
#define Euler() (0.57721566490153286060651209008240243104)
inline double round(double x) {return floor(x+0.5);}
inline void Iasb(long& a, double x) {a = (long)x;}
inline long longify(double x) {return (long)x;}
inline int doublify(const bigfloat& x, double& d) {d=x; return 0;}


inline double to_bigfloat(const int& n) {return double(n);}
inline double to_bigfloat(const long& n) {return double(n);}
inline double to_bigfloat(const double& x) {return x;}

inline bigfloat I2bigfloat(const bigint& x) {return I2double(x);}

// complexes

#include <complex>
#define bigcomplex complex<double>

inline int is_zero(const bigcomplex& z) 
 {return is_zero(z.real())&&is_zero(z.imag());}
inline int is_approx_zero(const bigcomplex& z)
 {return is_approx_zero(z.real())&&is_approx_zero(z.imag());}

//////////////////////////////////////////////////////////////////
#endif // NTL_ALL
//////////////////////////////////////////////////////////////////

#undef setbit

#endif // #define _INTERFACE_H_
