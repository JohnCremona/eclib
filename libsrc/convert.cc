// convert.h: declarations of integer conversion functions PARI/NTL/FLINT
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2025 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
// Conversions between multipreciaion integers in NTL, FLINT, PARI:

//#define DEBUG_CONVERT

#ifdef DEBUG_CONVERT
#include <iostream>
using std::cout;
using std::cerr;
using std::cin;
using std::endl;
#endif

#include "eclib/convert.h"
#include <NTL/ZZ_limbs.h>

using namespace NTL;

#if FLINT_LONG_LONG
typedef unsigned long long int ulong;
typedef long long int slong;
#else
typedef unsigned long int ulong;
typedef long int slong;
#endif

ZZ FLINT_to_NTL(const fmpz_t& a)  // from FLINT to NTL
{
  if (fmpz_fits_si(a))
    return ZZ(fmpz_get_si(a));

  // deal with sign (a is certainly nonzero now)
  int sign_a = fmpz_sgn(a);
  fmpz_t aa; fmpz_init(aa);
  fmpz_abs(aa, a);

  // extract limbs
  slong n = fmpz_size(aa);
  ulong* limbs = new ulong[n];
  fmpz_get_ui_array(limbs, n, aa);

  // construct ZZ from limbs
  ZZ b;
  ZZ_limbs_set(b, (ZZ_limb_t*)limbs, n);

  // return with sign
  return (sign_a<0? -b : b);
}

ZZ PARI_to_NTL(const GEN& a)  // from PARI to NTL
{
  // if a fits in a long int it is easy:
  if (!is_bigint(a))
    return ZZ(itos(a));

  // deal with sign (a is certainly nonzero now)
  int sign_a = signe(a);
  GEN aa = absi(a);

#ifdef DEBUG_CONVERT
  pari_printf("In NTL_to_PARI() with t_INT a =  %Ps, abs(a) = %Ps\n", a, aa);
#endif
  // extract digits as a t_VEC
  GEN digits = binary_2k(a, NTL_BITS_PER_LIMB_T);
#ifdef DEBUG_CONVERT
  cout << "NTL_BITS_PER_LIMB_T = " << NTL_BITS_PER_LIMB_T << endl;
  pari_printf("digits:  %Ps\n", digits);
#endif

  // convert this to an array of ZZ_limb_t
  int n = lg(digits)-1;
  ZZ_limb_t* limbs = new ZZ_limb_t[n];
  for (int i=1; i<lg(digits); i++)
    {
      limbs[n-i] = itou(gel(digits,i));
    }
#ifdef DEBUG_CONVERT
  cout << "Limbs for NTL: ";
  for (int i=0; i<n; i++)
    {
      if (i) cout << ", ";
      cout << limbs[i];
    }
  cout << endl;
#endif

  // construct ZZ from limbs
  ZZ b;
  ZZ_limbs_set(b, limbs, n);

  // return with sign
  return (sign_a<0? -b : b);
}

fmpz_t* NTL_to_FLINT(const ZZ& a)  // from NTL to FLINT
{
  if (IsZero(a))
    {
      fmpz_t* b = new fmpz_t[1];
      fmpz_init_set_si(b[0], 0);
      return b;
    }

  // deal with sign (now a is nonzero)
  int sign_a = sign(a);
  ZZ aa = abs(a);

#ifdef DEBUG_CONVERT
  cout<<"In NTL_to_FLINT() with ZZ a = " << a << ", abs(a) = " << aa <<endl;
#endif
  // extract limbs
  int n = aa.size();
  const ZZ_limb_t* limbs = ZZ_limbs_get(aa);
#ifdef DEBUG_CONVERT
  cout<<" #limbs = " << n << "\nlimbs: ";
  for(int i=0; i<n; i++)
    {
      if (i) cout << ", ";
      cout << limbs[i] << " ";
    }
  cout << endl;
#endif

  // construct an fmpz_t from these limbs:
  fmpz_t* b = new fmpz_t[1]; fmpz_init(b[0]);
  fmpz_set_ui_array(b[0], limbs, n);

  // adjust sign if necessary
  if (sign_a<0)
    fmpz_neg(b[0],b[0]);
  return b;
}

fmpz_t* PARI_to_FLINT(const GEN& a)  // from PARI to FLINT
{
  // if a fits in a long int it is easy:
  if (!is_bigint(a))
    {
      fmpz_t* b = new fmpz_t[1];
      fmpz_set_si(b[0], itos(a));
      return b;
    }
  // deal with sign (a is certainly nonzero here)
  int sign_a = signe(a);
  GEN aa = absi(a);

  // extract digits as a t_VEC
  GEN digits = binary_2k(aa, FLINT_BITS); // n's digits in base 2^64, as a t_VEC

  // convert this to an array of ulongs
  int n = lg(digits)-1;
  ulong* limbs = new ulong[n];
  for (int i=1; i<=n; i++)
    {
      limbs[n-i] = itou(gel(digits,i));
    }

  // construct an fmpz_t from these limbs:
  fmpz_t* b = new fmpz_t[1]; fmpz_init(b[0]);
  fmpz_set_ui_array(b[0], limbs, n);

  // adjust sign if necessary
  if (sign_a<0)
    fmpz_neg(b[0],b[0]);
  return b;
}

GEN NTL_to_PARI(const ZZ& a)  // from NTL to PARI
{
  if (IsZero(a))
    return stoi(0);

  // deal with sign (now a is nonzero)
  int sign_a = sign(a);
  ZZ aa = abs(a);

#ifdef DEBUG_CONVERT
  cout<<"In NTL_to_PARI() with ZZ a = " << a << ", abs(a) = " << aa <<endl;
#endif
  // extract limbs
  int n = aa.size();
  const ZZ_limb_t* limbs = ZZ_limbs_get(aa);

#ifdef DEBUG_CONVERT
  cout<<" #limbs = " << n << "\nlimbs:         ";
  for(int i=0; i<n; i++)
    {
      if (i) cout << ", ";
      cout << limbs[i] << " ";
    }
  cout << endl;
#endif

  // convert limbs array to a t_VEC
  GEN v = cgetg(n+1, t_VEC);
  for (int i=1; i<=n; i++)
    gel(v,i) = utoi(limbs[n-i]);

#ifdef DEBUG_CONVERT
    pari_printf("limbs as t_VEC: %Ps\n", v);
    cout << "NTL_BITS_PER_LIMB_T = " << NTL_BITS_PER_LIMB_T << endl;
#endif

  // construct t_INT from limbs
  GEN b = fromdigits_2k(v, NTL_BITS_PER_LIMB_T);
#ifdef DEBUG_CONVERT
  pari_printf("as t_INT: %Ps\n", b);
#endif

  // return with sign
  return (sign_a<0? negi(b): b);
}

GEN FLINT_to_PARI(const fmpz_t& a)  // from FLINT to PARI
{
  if (fmpz_fits_si(a))
    return stoi(fmpz_get_si(a));

  // deal with sign (a is certainly nonzero now)
  int sign_a = fmpz_sgn(a);

  fmpz_t aa; fmpz_init(aa);
  fmpz_abs(aa, a);

  // extract limbs
  slong n = fmpz_size(aa);
  ulong* limbs = new ulong[n];
  fmpz_get_ui_array(limbs, n, aa);

  // convert limbs array to a t_VEC
  GEN v = cgetg(n+1, t_VEC);
  for (int i=1; i<=n; i++)
    {
      gel(v,i) = utoi(limbs[n-i]);
    }

  // construct t_INT from limbs
  GEN b = fromdigits_2k(v, FLINT_BITS);

  // return with sign
  return (sign_a<0? negi(b): b);
}

