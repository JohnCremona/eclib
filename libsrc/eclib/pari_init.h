// pari_init.h: initialization of libpari
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2025 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

#ifndef _ECLIB_PARI_INIT_H
#define _ECLIB_PARI_INIT_H      1
                           //flags that this file has been included

#include <iostream>
namespace PARI{
  extern "C" {
#include <pari/pari.h>
  }
#undef coeff // since pari defines this as a macro with 3 args but NTL as a function with 2
#undef ulong // since pari defines this to be pari_ulong
}

using PARI::pari_sp;
using PARI::avma;
using PARI::itos;
using PARI::stoi;
using PARI::pari_printf;
using PARI::ellinit;
using PARI::ellap;
using PARI::Z_factor;
using PARI::nbrows;
using PARI::isprime;
using PARI::mkvecn;
using PARI::GEN;
using PARI::cgetg;
using PARI::itou;
using PARI::utoi;
using PARI::absi;
using PARI::negi;
using PARI::is_bigint;
using PARI::binary_2k;
using PARI::fromdigits_2k;
using PARI::pari_printf;
using PARI::t_VEC;
// NOT using PARI::signe, PARI::lg, PARI::gel which are macros

#define DEFAULT_PARI_SIZE 100000000
#define DEFAULT_PARI_MAX_PRIME 1000000

void eclib_pari_init(long max_prime=DEFAULT_PARI_MAX_PRIME);

#endif
