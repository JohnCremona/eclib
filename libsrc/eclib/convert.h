// convert.h: declarations of integer conversion functions PARI/NTL/FLINT
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2025 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#ifndef _ECLIB_CONVERT_H
#define _ECLIB_CONVERT_H

// Conversions between multipreciaion integers in NTL, FLINT, PARI:

#include <gmp.h>
#include <flint/flint.h> // must include this first to set __FLINT_VERSION
#include <flint/fmpz.h>
#include <NTL/ZZ.h>
#include "pari_init.h"

using NTL::ZZ;
using PARI::GEN;

ZZ FLINT_to_NTL  (const fmpz_t& a);   // from FLINT to NTL
ZZ PARI_to_NTL   (const GEN& a);      // from PARI to NTL
fmpz_t* NTL_to_FLINT  (const ZZ& a);  // from NTL to FLINT
fmpz_t* PARI_to_FLINT (const GEN& a); // from PARI to FLINT
GEN NTL_to_PARI   (const ZZ& a);      // from NTL to PARI
GEN FLINT_to_PARI (const fmpz_t& a);  // from FLINT to PARI

#endif

// end of file convert.h
