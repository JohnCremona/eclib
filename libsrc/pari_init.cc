// pari_init.cc: initialization of libpari
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2025 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

#include <iostream>

#include "eclib/interface.h"
#include "eclib/pari_init.h"       // must be included after any NTL

void eclib_pari_init(long max_prime)
{
  if (!avma) {
    long pari_size = strtol(getenv_with_default("PARI_SIZE", "DEFAULT_PARI_SIZE").c_str(), NULL, 0);
    if (pari_size==0) // e.g. syntax error in the environment variable PARI_SIZE
      pari_size = DEFAULT_PARI_SIZE;
    // the first parameter is the maximum stack size in bytes
    // the second parameter is the maximum precomputed prime
    pari_init(pari_size, max_prime);
  }
}
