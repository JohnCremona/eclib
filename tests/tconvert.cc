// tconvert.cc -- test program for integer conversion functions
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2025 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include <cassert>
#include <eclib/interface.h>
#include <eclib/convert.h>
#include <eclib/pari_init.h>

std::ostream& operator<<(std::ostream& s, const fmpz_t& a);

// Compiler cannot distinguish between fmpz_t (aka 'long int[1]') and GEN (aka 'long int*')
//std::ostream& operator<<(std::ostream& s, const GEN& a);

int test(const ZZ& a);

int main()
{
  eclib_pari_init();

  cout << "Testing conversions between types ZZ (NTL integers), fmpz_t (FLINT integers) and t_INT (PARI integers)" << endl;

  ZZ a;
  // cout << "Enter an integer a: "; cin >> a; cout << endl;
  // if (test(a))
  //   cout << "a = " << a << " PASSED" << endl << endl;
  // else
  //   cout << "a = " << a << " FAILED" << endl << endl;

  ZZ a0 = to_ZZ(0);
  ZZ a1 = to_ZZ("12345");
  ZZ a2 = to_ZZ("-98765");
  ZZ a3 = to_ZZ("-8472893749012374104710000000000000000000000000000000000001");
  ZZ a4 = to_ZZ("83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851");

  for (auto ai: {a0,a1,a2,a3,a4})
    if (test(ai))
      cout << "a = " << ai << " PASSED" << endl << endl;
    else
      cout << "a = " << ai << " FAILED" << endl << endl;
}

int test(const ZZ& a)
{
  cout << "Conversion tests" << endl;
  cout << "a = " << a << " (as ZZ)" << endl;

  fmpz_t* b = NTL_to_FLINT(a);
  cout << "a = " << b[0] << " (as fmpz_t)" << endl;
  ZZ a2 = FLINT_to_NTL(b[0]);
  cout << "converting back to ZZ...\na = " << a2 << endl;
  if (a==a2)
    cout << "OK" <<endl;
  else
    {
      cout << "******************* WRONG! ******************" <<endl;
      return 0;
    }
  GEN c = NTL_to_PARI(a);
  pari_printf("a = %Ps (as t_INT)\n", c);
  ZZ a3 = PARI_to_NTL(c);
  cout << "converting back to ZZ...\na = " << a3 << endl;
  if (a==a3)
    cout << "OK" <<endl;
  else
    {
      cout << "******************* WRONG! ******************" <<endl;
      return 0;
    }

  cout << "Converting from FLINT to PARI\n";
  GEN d = FLINT_to_PARI(b[0]);
  pari_printf("a = %Ps (as t_INT)\n", d);
  cout << "Converting from PARI back to FLINT\n";
  fmpz_t* e = PARI_to_FLINT(d);
  cout << "a = " << e[0] << " (as fmpz_t)" << endl;
  cout << "and then back from FLINT to NTL\n";
  ZZ f = FLINT_to_NTL(e[0]);
  cout << "a = " << f << " (as ZZ)" << endl;
  if (a==f)
    cout << "OK" <<endl;
  else
  {
      cout << "******************* WRONG! ******************" <<endl;
      return 0;
    }
  return 1;
}

std::ostream& operator<<(std::ostream& s, const fmpz_t& a)
{
  char* st = fmpz_get_str(NULL, 10, a);
  s << std::string(st);
  flint_free(st);
  return s;
}

// std::ostream& operator<<(std::ostream& s, const GEN& a)
// {
//   char* st = pari_sprintf("%Ps", a);
//   s << std::string(st);
//   free(st);
//   return s;
// }
