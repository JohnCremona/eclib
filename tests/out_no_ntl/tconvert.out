Testing conversions between types ZZ (NTL integers), fmpz_t (FLINT integers) and t_INT (PARI integers)
Conversion tests
a = 0 (as ZZ)
a = 0 (as fmpz_t)
converting back to ZZ...
a = 0
OK
a = 0 (as t_INT)
converting back to ZZ...
a = 0
OK
Converting from FLINT to PARI
a = 0 (as t_INT)
Converting from PARI back to FLINT
a = 0 (as fmpz_t)
and then back from FLINT to NTL
a = 0 (as ZZ)
OK
a = 0 PASSED

Conversion tests
a = 12345 (as ZZ)
a = 12345 (as fmpz_t)
converting back to ZZ...
a = 12345
OK
a = 12345 (as t_INT)
converting back to ZZ...
a = 12345
OK
Converting from FLINT to PARI
a = 12345 (as t_INT)
Converting from PARI back to FLINT
a = 12345 (as fmpz_t)
and then back from FLINT to NTL
a = 12345 (as ZZ)
OK
a = 12345 PASSED

Conversion tests
a = -98765 (as ZZ)
a = -98765 (as fmpz_t)
converting back to ZZ...
a = -98765
OK
a = -98765 (as t_INT)
converting back to ZZ...
a = -98765
OK
Converting from FLINT to PARI
a = -98765 (as t_INT)
Converting from PARI back to FLINT
a = -98765 (as fmpz_t)
and then back from FLINT to NTL
a = -98765 (as ZZ)
OK
a = -98765 PASSED

Conversion tests
a = -8472893749012374104710000000000000000000000000000000000001 (as ZZ)
a = -8472893749012374104710000000000000000000000000000000000001 (as fmpz_t)
converting back to ZZ...
a = -8472893749012374104710000000000000000000000000000000000001
OK
a = -8472893749012374104710000000000000000000000000000000000001 (as t_INT)
converting back to ZZ...
a = -8472893749012374104710000000000000000000000000000000000001
OK
Converting from FLINT to PARI
a = -8472893749012374104710000000000000000000000000000000000001 (as t_INT)
Converting from PARI back to FLINT
a = -8472893749012374104710000000000000000000000000000000000001 (as fmpz_t)
and then back from FLINT to NTL
a = -8472893749012374104710000000000000000000000000000000000001 (as ZZ)
OK
a = -8472893749012374104710000000000000000000000000000000000001 PASSED

Conversion tests
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851 (as ZZ)
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851 (as fmpz_t)
converting back to ZZ...
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851
OK
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851 (as t_INT)
converting back to ZZ...
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851
OK
Converting from FLINT to PARI
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851 (as t_INT)
Converting from PARI back to FLINT
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851 (as fmpz_t)
and then back from FLINT to NTL
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851 (as ZZ)
OK
a = 83475623875628564398568356325856876198561566179865781346578165843561854643856198564189564184651856148356184654187561856148561856143856431851 PASSED

